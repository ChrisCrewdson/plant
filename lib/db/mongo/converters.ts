import { ObjectID } from 'mongodb';
import utils from '../../../app/libs/utils';

export const convertPlantDataTypesForSaving = (
  plantIn: Readonly<BizPlant> | Readonly<UiPlantsValue>):
  Readonly<DbPlant> => {
  const plant: DbPlant = plantIn as unknown as DbPlant;
  if (plant._id) {
    plant._id = new ObjectID(plant._id);
  }
  if (!plant.userId) {
    throw new Error('Missing userId in convertPlantDataTypesForSaving');
  }
  if (!plant.locationId) {
    throw new Error('Missing locationId in convertPlantDataTypesForSaving');
  }
  plant.userId = new ObjectID(plant.userId);
  plant.locationId = new ObjectID(plant.locationId);

  if (plant.plantedDate) {
    plant.plantedDate = utils.dateToInt(plant.plantedDate);
  }
  if (plant.purchasedDate) {
    plant.purchasedDate = utils.dateToInt(plant.purchasedDate);
  }
  if (plant.terminatedDate) {
    plant.terminatedDate = utils.dateToInt(plant.terminatedDate);
  }

  return plant;
};
