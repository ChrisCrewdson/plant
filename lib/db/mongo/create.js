const Helper = require('./helper');

module.exports = class Create {
  /**
   * Create one or more documents in collection
   * @param {import('mongodb').Db} db
   * @param {string} collection
   * @param {object} doc
   */
  static async create(db, collection, doc) {
    const coll = db.collection(collection);
    const body = await coll.insert(Helper.removeEmpty(doc));
    return body && body.ops;
  }

  /**
   * Create single document in collection
   * @param {import('mongodb').Db} db
   * @param {string} collection
   * @param {object} doc
   */
  static async createOne(db, collection, doc) {
    const body = await Create.create(db, collection, doc);
    return body && body[0];
  }
};
