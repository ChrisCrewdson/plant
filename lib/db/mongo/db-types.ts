import { DbNote } from './model-note';

interface GoogleOAuthJson {
  id: string;
}

export interface GoogleOAuth {
  _json: GoogleOAuthJson;
  displayName: string;
  /**
   * Objects look like: { value: string ==> email address }
   */
  emails: object[];
}

interface FacebookOAuthJson {
  emails: string[]|string;
  id: string;
}

interface FacebookOAuthName {
  givenName: string;
  familyName: string;
}

export interface FacebookOAuth {
  _json: FacebookOAuthJson;
  name: FacebookOAuthName;
}

// Generic definitions that apply to multiple models.
// For specific models see the model-<name>.d.ts file.

// The model-<name>.d.ts files should follow a specific pattern where the shapes
// for the DB are defined at the top, then the Biz layer and then the Ui layer.
// Each shape will build on and change the previous shape.

export type DbShape = DbLocation | DbNote | DbPlant | DbUser | DbUserTiny;

export type DbShapes = DbLocation[] | DbNote[] | DbPlant[] | DbUser[] | DbUserTiny[];

export type DbCollectionName = 'user' | 'plant' | 'note' | 'location';

export interface UserDetails {
  // TODO: There will only ever be facebook or google set here so this show be a union
  // type such as:
  // socialJson: FacebookOAuthJson|GoogleOAuthJson;
  facebook?: FacebookOAuthJson;
  google?: GoogleOAuthJson;
  name?: string;
  createdAt?: Date;
  updatedAt?: Date;
  email?: string;
}
