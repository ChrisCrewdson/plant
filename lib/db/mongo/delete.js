
/**
 * Delete from collection
 * @param {import('mongodb').Db} db
 * @param {DbCollectionName} collection
 * @param {object} query
 * @returns {Promise<number|undefined>}
 */
module.exports = async (db, collection, query) => {
  const coll = db.collection(collection);
  const results = await coll.deleteMany(query);
  // results if query item was found:
  //
  // { result: { ok: 1, n: 3 }, deletedCount: 3
  //
  // result if query item was not found:
  //
  // { result: { ok: 1, n: 0 }, deletedCount: 0
  return results && results.deletedCount;
};
