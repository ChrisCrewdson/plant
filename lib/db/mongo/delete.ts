
export const remove = async (db: import('mongodb').Db, collection: DbCollectionName, query: object):
 Promise<number | undefined> => {
  const coll = db.collection(collection);
  const results = await coll.deleteMany(query);
  // results if query item was found:
  //
  // { result: { ok: 1, n: 3 }, deletedCount: 3
  //
  // result if query item was not found:
  //
  // { result: { ok: 1, n: 0 }, deletedCount: 0
  return results && results.deletedCount;
};
