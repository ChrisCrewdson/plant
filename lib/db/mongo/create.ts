import { Db } from 'mongodb';
import { Helper } from './helper';
import { DbCollectionName, UserDetails, DbShape } from './db-types';

// See https://stackoverflow.com/a/36794912/1463
// for a description of insertOne/insertMany/bulkWrite

export class Create {
  static async create(db: Db, collection: DbCollectionName, doc: object):
  Promise<any> {
    const coll = db.collection(collection);
    const body = await coll.insertOne(Helper.removeEmpty(doc));
    return body && body.ops;
  }

  // TODO: Fix - remove UserDetails
  static async createOne(db: Db, collection: DbCollectionName, doc: DbShape | UserDetails):
   Promise<any> {
    const body = await Create.create(db, collection, doc);
    return body && body[0];
  }

  static async createLocation(db: Db, doc: DbLocation):
   Promise<DbLocation> {
    return Create.createOne(db, 'location', doc);
  }

  static async createNote(db: Db, doc: DbNote): Promise<DbNote> {
    return Create.createOne(db, 'note', doc);
  }

  static async createPlant(db: Db, doc: DbPlant): Promise<DbPlant> {
    return Create.createOne(db, 'plant', doc);
  }

  static async createUser(db: Db, doc: UserDetails): Promise<DbUser> {
    return Create.createOne(db, 'user', doc);
  }
}
