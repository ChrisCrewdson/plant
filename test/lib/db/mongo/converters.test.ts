import { convertPlantDataTypesForSaving } from '../../../../lib/db/mongo/converters';

describe('/lib/db/mongo/converters', () => {
  describe('convertPlantDataTypesForSaving', () => {
    test('should do basic conversion', () => {
      const plantIn: BizPlant = {
        _id: '5db7bc3796e0526dfa41c6e0',
        locationId: '5db7bc3796e0526dfa41c6e0',
        userId: '5db7bc3796e0526dfa41c6e0',
        title: 'Plant Title',
      };
      const result = convertPlantDataTypesForSaving(plantIn);
      expect(result).toMatchSnapshot();
    });

    test('should throw if no userId', () => {
      const plantIn = {
        _id: '5db7bc3796e0526dfa41c6e0',
        locationId: '5db7bc3796e0526dfa41c6e0',
        title: 'Plant Title',
      } as BizPlant;
      expect(() => convertPlantDataTypesForSaving(plantIn))
        .toThrowErrorMatchingSnapshot();
    });

    test('should throw if no locationId', () => {
      const plantIn = {
        _id: '5db7bc3796e0526dfa41c6e0',
        userId: '5db7bc3796e0526dfa41c6e0',
        title: 'Plant Title',
      } as BizPlant;
      expect(() => convertPlantDataTypesForSaving(plantIn))
        .toThrowErrorMatchingSnapshot();
    });

    test('should convert dates as well', () => {
      const plantIn: UiPlantsValue = {
        _id: '5db7bc3796e0526dfa41c6e0',
        locationId: '5db7bc3796e0526dfa41c6e0',
        userId: '5db7bc3796e0526dfa41c6e0',
        title: 'Plant Title',
        purchasedDate: '12/15/2019',
        terminatedDate: '02/29/2020',
        plantedDate: '01/01/2019',
      };
      const result = convertPlantDataTypesForSaving(plantIn);
      expect(result).toMatchSnapshot();
    });
  });
});
