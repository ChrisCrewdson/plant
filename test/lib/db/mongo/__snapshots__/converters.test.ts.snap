// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/lib/db/mongo/converters convertPlantDataTypesForSaving should convert dates as well 1`] = `
Object {
  "_id": "5db7bc3796e0526dfa41c6e0",
  "locationId": "5db7bc3796e0526dfa41c6e0",
  "plantedDate": 20190101,
  "purchasedDate": 20191215,
  "terminatedDate": 20200229,
  "title": "Plant Title",
  "userId": "5db7bc3796e0526dfa41c6e0",
}
`;

exports[`/lib/db/mongo/converters convertPlantDataTypesForSaving should do basic conversion 1`] = `
Object {
  "_id": "5db7bc3796e0526dfa41c6e0",
  "locationId": "5db7bc3796e0526dfa41c6e0",
  "title": "Plant Title",
  "userId": "5db7bc3796e0526dfa41c6e0",
}
`;

exports[`/lib/db/mongo/converters convertPlantDataTypesForSaving should throw if no locationId 1`] = `"Missing locationId in convertPlantDataTypesForSaving"`;

exports[`/lib/db/mongo/converters convertPlantDataTypesForSaving should throw if no userId 1`] = `"Missing userId in convertPlantDataTypesForSaving"`;
